// @ark/schema/out/structure/required.js
import { unset } from "@ark/util";
import { compileErrorContext, implementNode } from "../shared/implement.js";
import { BaseProp, intersectProps } from "./prop.js";
const implementation = implementNode({
    kind: "required",
    hasAssociatedError: true,
    intersectionIsOpen: true,
    keys: {
        key: {},
        value: {
            child: true,
            parse: (schema, ctx) => ctx.$.parseSchema(schema)
        }
    },
    normalize: schema => schema,
    reduce: (inner, $) => {
        if (inner.value.defaultMeta !== unset) {
            return $.node("optional", {
                ...inner,
                default: inner.value.defaultMeta
            });
        }
        if (inner.value.optionalMeta)
            return $.node("optional", inner);
    },
    defaults: {
        description: node => `${node.compiledKey}: ${node.value.description}`,
        expected: ctx => ctx.missingValueDescription,
        actual: () => "missing"
    },
    intersections: {
        required: intersectProps,
        optional: intersectProps
    }
});
export class RequiredNode extends BaseProp {
    expression = `${this.compiledKey}: ${this.value.expression}`;
    errorContext = Object.freeze({
        code: "required",
        missingValueDescription: this.value.shortDescription,
        relativePath: [this.key]
    });
    compiledErrorContext = compileErrorContext(this.errorContext);
}
export const Required = {
    implementation,
    Node: RequiredNode
};
